backend-rust/src/graphql_api.rs

```diff
@@ -1313,14 +1331,209 @@ struct Contract {
+    // This function returns events from the `contract_events` table as well as
+    // one `init_transaction_event` from when the contract was initialized. The
+    // `skip` and `take` parameters are used to paginate the events.
     async fn contract_events(
         &self,
+        ctx: &Context<'_>,
         skip: i32,
         take: i32,
     ) -> ApiResult<ContractEventsCollectionSegment> {
-        todo_api!()
+        let config = get_config(ctx)?;
+        let pool = get_pool(ctx)?;
+
+        // If `skip` is 0 and at least one event is taken, include the
+        // `init_transaction_event`.
+        let include_initial_event = skip == 0 && take > 0;
+        // Adjust the `take` and `skip` values considering if the
+        // `init_transaction_event` is requested to be included or not.
+        let take_without_initial_event =
+            std::cmp::max(0, take as i64 - include_initial_event as i64);
+        let skip_without_initial_event = std::cmp::max(0, skip as i64 - 1i64);
+
+        // Limit the number of events to be fetched from the `contract_events` table.
+        let limit = std::cmp::min(
+            take_without_initial_event,
+            std::cmp::max(
+                0,
+                config.contract_events_collection_limit - include_initial_event as i64,
+            ),
+        );
+
+        let mut contract_events = vec![];
+        let mut total_events_count = 0;
+
+        // Get the events from the `contract_events` table.
+        let row_stream = sqlx::query!(
+            r#"
+                            SELECT * FROM (
+                                SELECT
+                                    contract_events.index,
+                                    contract_events.transaction_index,
+                                    trace_element_index,
+                                    contract_events.block_height AS event_block_height,
+                                    transactions.hash as transaction_hash,
+                                    transactions.events,
+                                    accounts.address as creator,
+                                    blocks.slot_time as block_slot_time,
+                                    blocks.height as block_height
+                                FROM contract_events
+                                JOIN transactions
+                                ON contract_events.block_height = transactions.block_height
+                                AND contract_events.transaction_index = transactions.index
+                                JOIN accounts
+                                ON transactions.sender = accounts.index
+                                JOIN blocks ON contract_events.block_height = blocks.height
+                                WHERE contract_events.contract_index = $1
+                                AND contract_events.contract_sub_index <= $2
+                                LIMIT $3
+                                OFFSET $4
```

Comment:
Yes, that would be a way to solve the inefficiency. Considering the sequential way we insert stuff in the indexer, this is honestly probably the best solution, since it's quite easy to add that index column.